ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ImmutableDB/StateMachine.hs

```diff
@@ -772,39 +781,59 @@ postcondition model cmdErr resp =
   where
     ev = lockstep model cmdErr resp
 
+data ImmutableDBState = ImmutableDBState
+    { db       :: ImmutableDB    Hash IO
+    , internal :: ImmDB.Internal Hash IO
+    }
+  deriving NoUnexpectedThunks via AllowThunk ImmutableDBState
+
 -- | Environment to run commands against the real ImmutableDB implementation.
 data ImmutableDBEnv h = ImmutableDBEnv
-  { varErrors  :: StrictTVar IO Errors
-  , varNextId  :: StrictTVar IO Id
-  , varCurSlot :: StrictTVar IO SlotNo
+  { varErrors   :: StrictTVar IO Errors
+  , varNextId   :: StrictTVar IO Id
+  , varCurSlot  :: StrictTVar IO SlotNo
     -- ^ Always 'maxBound'. Only when executing 'ReopenInThePast', it will
     -- temporarily be reset to a slot in the past.
-  , varIters   :: StrictTVar IO [TestIterator IO]
+  , varIters    :: StrictTVar IO [TestIterator IO]
     -- ^ A list of all open iterators. For some commands, e.g., corrupting the
     -- database or simulating errors, we need to close and reopen the
     -- database, which almost always requires truncation of the database.
     -- During truncation we might need to delete a file that is still opened
     -- by an iterator. As this is not allowed by the MockFS implementation, we
     -- first close all open iterators in these cases.
-  , registry   :: ResourceRegistry IO
-  , hasFS      :: HasFS IO h
-  , db         :: ImmutableDB    Hash IO
-  , internal   :: ImmDB.Internal Hash IO
+  , registry    :: ResourceRegistry IO
+  , hasFS       :: HasFS IO h
+  , varDB       :: StrictMVar IO ImmutableDBState
+    -- The fields below are only used to reopen the ImmutableDB
+  , chunkInfo   :: ChunkInfo
+  , parser      :: ChunkFileParser
+                     (ChunkFileError TestHeaderHash)
+                     IO
+                     (BlockSummary TestHeaderHash)
+                     TestHeaderHash
+  , tracer      :: Tracer IO (TraceEvent (ChunkFileError TestHeaderHash) TestHeaderHash)
+  , cacheConfig :: Index.CacheConfig
+  , btime       :: BlockchainTime IO
```

Comment:
Introduce `ImmutableDbArgs` (keep the parser separate) in the `ImmutableDB`. Do the same for the VolatileDB.